src/CompHash.cc

```diff
@@ -359,10 +359,11 @@ bool CompositeHash::RecoverOneVal(const HashKey& hk, Type* t, ValPtr* pval, bool
 
 					ASSERT(int(values.size()) == num_fields);
 
-					auto rv = make_intrusive<RecordVal>(IntrusivePtr{NewRef{}, rt});
+					auto rv = make_intrusive<RecordVal>(IntrusivePtr{NewRef{}, rt},
+					                                    false /* init_fields */);
 
 					for ( int i = 0; i < num_fields; ++i )
-						rv->Assign(i, std::move(values[i]));
+						rv->AppendField(std::move(values[i]), rt->GetFieldType(i));
```

Comment:
If the field is optional, does this work okay because `values[i]` will be nil?

Reply:
Yes, values[i] will be nil for unset values. This is modeled after `RecordVal::DoClone()`.